compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -275,6 +295,36 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
     ) -> QueryResult<'tcx> {
         self.make_canonical_response(Certainty::Yes)
     }
+
+    fn compute_subtype_goal(
+        &mut self,
+        goal: Goal<'tcx, SubtypePredicate<'tcx>>,
+    ) -> QueryResult<'tcx> {
+        self.infcx.probe(|_| {
+            let InferOk { value: (), obligations } = self
+                .infcx
+                .at(&ObligationCause::dummy(), goal.param_env)
+                .sub(goal.predicate.a, goal.predicate.b)?;
+            self.evaluate_all_and_make_canonical_response(
+                obligations.into_iter().map(|pred| pred.into()).collect(),
+            )
+        })
+    }
+
+    fn compute_closure_kind_goal(
+        &mut self,
+        found_kind: Option<ty::ClosureKind>,
+        expected_kind: ty::ClosureKind,
+    ) -> QueryResult<'tcx> {
+        let Some(found_kind) = found_kind else {
+            return self.make_canonical_response(Certainty::Maybe(MaybeCause::Ambiguity));
```

Comment:
Should I make a "make dummy ambiguous response" function on `ecx`? 

I feel like there are a lot of places where we're returning a canonical response with no constraints, but we've already entered the canonical goal, so we can't call `response_no_constraints`.

Reply:
Also, should I make `Certainty::Maybe(MaybeCause::Ambiguity)` into a constant... `Certainty::AMBIGUOUS`? I've been typing the former a lot.

Reply:
> Also, should I make `Certainty::Maybe(MaybeCause::Ambiguity)` into a constant... `Certainty::AMBIGUOUS`? I've been typing the former a lot.

yeah, seems good to me. I think as a user you pretty much always want to treat `Certainty::Maybe(_)` in the same way but the constant does feel useful when creating the response.

> Should I make a "make dummy ambiguous response" function on `ecx`? 
>
> I feel like there are a lot of places where we're returning a canonical response with no constraints, but we've already entered the canonical goal, so we can't call `response_no_constraints`.

I guess that would be helpful for perf :shrug: don't have a strong opinion here, could wait until the first perf run.